oprf/group/group.go

```diff
@@ -0,0 +1,188 @@
+package group
+
+import (
+	"errors"
+	"math/big"
+)
+
+// Point is a representation of a group element. It has two coordinates of
+// *big.Int type.
+type Point struct {
+	c *Ciphersuite
+	x *big.Int
+	y *big.Int
+}
+
+// Element is the interface that represents points in a given
+// Ciphersuite instantiation.
+type Element interface {
+	// Returns a bool indicating that the Point is valid
+	IsValid() bool
+
+	// Returns a bool indicating that the Point is the identity
+	IsIdentity() bool
+
+	// Performs a scalar multiplication of the Point with some scalar
+	// input
+	ScalarMult(*Scalar) (*Point, error)
+
+	// Performs the addition operation on the calling Point object
+	// along with a separate Point provided as input
+	Add(*Point) (*Point, error)
+
+	// Performs the negation operation on the calling Point object
+	Neg() (*Point, error)
+
+	// Serializes the Point into a byte slice
+	Serialize() ([]byte, error)
+
+	// Attempts to deserialize a byte slice into a Point
+	Deserialize([]byte) error
+
+	// Returns a bool indicating whether two Points are equal
+	Equal(*Point) bool
+}
+
+// NewPoint generates a new point.
+func NewPoint(c *Ciphersuite) *Point {
+	p := &Point{c, new(big.Int), new(big.Int)}
+	return p
+}
+
+// IsValid checks that the given Point is a valid curve point
+func (p *Point) IsValid() bool {
+	return p.c.curve.IsOnCurve(p.x, p.y)
+}
+
+// ScalarMult multiplies a Point by the provided Scalar value
+func (p *Point) ScalarMult(s *Scalar) (*Point, error) {
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	q := NewPoint(p.c)
+	q.x, q.y = p.c.curve.ScalarMult(p.x, p.y, s.Serialize())
+
+	return q, nil
+}
+
+// Add performs the addition operation on the calling Point object
+// along with a separate Point provided as input
+func (p *Point) Add(q *Point) (*Point, error) {
+	if !p.IsValid() || !q.IsValid() {
```

Comment:
No need to check whether point is valid, since every instance of ~point~ element must be valid all the time.

Reply:
Changed ;)